service/storage_service.cc

```diff
@@ -561,6 +566,66 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
 
         bool res;
         switch (node.rs->ring.value().state) {
+            case ring_slice::replication_state::commit_cdc_generation: {
+                // make sure all nodes know about new topology and have the new CDC generation data
+                // (we require all nodes to be alive for topo change for now)
```

Comment:
I don't know. We'll have to go back to the drawing board.

CDC generations management currently assumes that all nodes participate in topology changes - but so does token-ring-based replication. We can ignore a node that's dead, if the admin tells us that the node is really dead and not partitioned away, sure, but that's it.

If we drop the assumption with tablets, CDC design will have to change, but it will have to change for tablet partitioning anyway.

Reply:
> I don't know. We'll have to go back to the drawing board.

We have the drawing board in the form of my document. We need to amend it.

> 
> CDC generations management currently assumes that all nodes participate in topology changes - but so does token-ring-based replication. We can ignore a node that's dead, if the admin tells us that the node is really dead and not partitioned away, sure, but that's it.

We want to allow "non affected" nodes to be able to fail during topology op. This will allow proceed with a topology change even if a node in different DC dies or an entire DC got partitioned.

Reply:
> We want to allow "non affected" nodes to be able to fail during topology op.

Ok but with token-ring-based replication, all nodes are affected on any topology change. One of the reasons is that any node might become a coordinator to any replica set, even outside that set. And CDC piggy-backs on this. CDC generation management requires updating all potential coordinators consistently across the cluster - which means all nodes.

With tablets it may be that only some small subset of nodes is affected during tablet movement. Combine this with fencing requests from outdated-coordinators, and we won't have to ensure that all coordinators are updated in time - just the replicas that are affected.

Reply:
> Affected nodes are those that gain or lose a vnode range

A node may stop or start being a replica for some moving range not only because it owns this range, but because one of its ranges is near the moving range in the token ring (e.g. it's within the next 3 tokens in the ring with RF=3). With 256 tokens and clusters, say, up to 100 nodes, with RF=3, we can be pretty sure that for *every* node, there is at least one range for which it stops or starts being a replica. So with today's topology changes which insert or remove 256 tokens to/from the token ring, while in theory not all nodes have to be affected, in practice all are.

> The question is how CDC fits in this model.

The entire set of CDC streams is changing on a generation switch, and all coordinators must learn about the new streams or they will coordinate to the wrong (old) streams which is bad if those writes are successful.

Reply:
> > Affected nodes are those that gain or lose a vnode range
> 
> A node may stop or start being a replica for some moving range not only because it owns this range, but because one of its ranges is near the moving range in the token ring (e.g. it's within the next 3 tokens in the ring with RF=3). With 256 tokens and clusters, say, up to 100 nodes, with RF=3, we can be pretty sure that for _every_ node, there is at least one range for which it stops or starts being a replica. So with today's topology changes which insert or remove 256 tokens to/from the token ring, while in theory not all nodes have to be affected, in practice all are.

For a single DC you are probably right. But for multiple DCs only nodes in a DC where operation is going on will be affected (according to our definition). 

> 
> > The question is how CDC fits in this model.
> 
> The entire set of CDC streams is changing on a generation switch, and all coordinators must learn about the new streams or they will coordinate to the wrong (old) streams which is bad if those writes are successful.

This is true for regular write coordinators as well, but we solve the problem by fencing their requests after draining all coordinators. If a non affected coordinator failed to be drained (because it was unreachable from the topology change coordinator)   it new requests will fail. Can the same be done for CDC?

Reply:
> For a single DC you are probably right. But for multiple DCs only nodes in a DC where operation is going on will be affected (according to our definition).

If there are zero SimpleStrategy/EverywhereStrategy keyspaces, maybe.

> This is true for regular write coordinators as well, but we solve the problem by fencing their requests after draining all coordinators. If a non affected coordinator failed to be drained (because it was unreachable from the topology change coordinator) it new requests will fail. Can the same be done for CDC?

Well I don't see how that's supposed to work for changing CDC streams. Sure the unreachable coordinator's requests will fail if the only nodes it tries to contact to complete the requests, are nodes that changed their topology version. But it might complete its requests using nodes that were unreachable during the topology change, including itself. These requests will be successful and they'll go to the wrong streams.

In general:
- if there's at least one node that was unreachable during CDC generation switch, that node will be able to successfully complete CL=ONE writes using the wrong streams
- if there are at least two nodes that were unreachable during CDC generation switch, those nodes will be able to successfully complete CL=TWO (CL=QUORUM with RF=3 etc.) writes using the wrong streams
- and so on.

That's why I said if we want to do topology changes without all nodes participating, we'll need to go back to the drawing board. Not for topology coordinator. I meant drawing board for redesigning how we change CDC streams completely from scratch.

Reply:
> > For a single DC you are probably right. But for multiple DCs only nodes in a DC where operation is going on will be affected (according to our definition).
> 
> If there are zero SimpleStrategy/EverywhereStrategy keyspaces, maybe.

Everywhere strategy should not matter I think. We just send everywhere anyway. SimplyStrategy yes. But this is normal requirement. 

> 
> > This is true for regular write coordinators as well, but we solve the problem by fencing their requests after draining all coordinators. If a non affected coordinator failed to be drained (because it was unreachable from the topology change coordinator) it new requests will fail. Can the same be done for CDC?
> 
> Well I don't see how that's supposed to work for changing CDC streams. Sure the unreachable coordinator's requests will fail if the only nodes it tries to contact to complete the requests, are nodes that changed their topology version. But it might complete its requests using nodes that were unreachable during the topology change, including itself. These requests will be successful and they'll go to the wrong streams.

Those will not be affected by the topology change and hence have compatible topology. If CDC uses the same set of replicas as a regular write  the same principal should apply. Otherwise "affected nodes" should be expanded.